Expose a 'category' for Web Preferences

brentfulgham opened the pull request:
#### 176591df93a53b8c551157e6c735e8e24d3fd1c1
<pre>
Expose a &apos;category&apos; for Web Preferences
<a href="https://bugs.webkit.org/show_bug.cgi?id=252093">https://bugs.webkit.org/show_bug.cgi?id=252093</a>
&lt;rdar://104962882&gt;

Reviewed by Patrick Angle.

Add a new &apos;category&apos; field to our preference definitions that can be used to provide context
to the User Interface for applications like Safari so that we can easily present the features
in the clearest way.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Scripts/GenerateSettings.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/APIFeature.cpp
(API::Feature::create): Add new argument.
(API::Feature::Feature): Ditto.
* Source/WebKit/UIProcess/API/APIFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm:
(-[_WKFeature category]):
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.mm.erb:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesInternalFeatures.mm.erb:
* Source/WebKitLegacy/mac/WebView/WebFeature.h:
* Source/WebKitLegacy/mac/WebView/WebFeature.m:
(-[WebFeature initWithKey:preferenceKey:name:status:category:details:defaultValue:hidden:]): Added.
(-[WebFeature initWithKey:preferenceKey:name:status:details:defaultValue:hidden:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h:

Canonical link: <a href="https://commits.webkit.org/260420@main">https://commits.webkit.org/260420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6105221682247f480203441a018e98b8153fc096

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108242 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117364 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112129 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8611 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100455 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114011 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97304 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42023 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/111776 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28946 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97445 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10180 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30291 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8291 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7196 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16329 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49887 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105832 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7209 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12499 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26199 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKitLegacy/mac/WebView/WebFeature.h:
```diff
@@ -50,12 +50,26 @@ typedef NS_ENUM(NSUInteger, WebFeatureStatus) {
 
+/*!
+ @enum WebFeatureCategory
+ @abstract Field indicating the purpose and level of completeness of a web feature. Used to determine which UI (if any) should reveal a feature.
```
Oops: This comment is wrong.

emw-apple commented on Source/WebKit/UIProcess/API/APIFeature.cpp:
```diff
@@ -28,16 +28,17 @@
+Feature::Feature(const WTF::String& name, const WTF::String& key, FeatureStatus status, FeatureCategory category, const WTF::String& details, bool defaultValue, bool hidden)
     : m_name(name)
     , m_key(key)
-    , m_status(status)
```
nit: Are these in the right order? I see that they don’t match the order of the constructor parameters.

brentfulgham replied:
These are in the correct initialization order to avoid a compile error. I adjusted the order slightly to group the small date types to improve packing.

emw-apple commented on Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesInternalFeatures.mm.erb:
```diff
@@ -43,7 +43,7 @@ using namespace WebKit;
 #if <%= @pref.condition %>
 <%- end -%>
-        adoptNS([[WebFeature alloc] initWithKey:@"<%= @pref.name %>" preferenceKey:@"<%= @pref.preferenceKey %>" name:@<%= @pref.humanReadableName %> status:<%= @pref.webFeatureStatus %> details:@<%= @pref.humanReadableDescription %> defaultValue:DEFAULT_VALUE_FOR_<%= @pref.name %> hidden:<%= @pref.hidden %>]).get(),
+        adoptNS([[WebFeature alloc] initWithKey:@"<%= @pref.name %>" preferenceKey:@"<%= @pref.preferenceKey %>" name:@<%= @pref.humanReadableName %> status:<%= @pref.webFeatureStatus %> category:WebFeatureCategoryNone details:@<%= @pref.humanReadableDescription %> defaultValue:DEFAULT_VALUE_FOR_<%= @pref.name %> hidden:<%= @pref.hidden %>]).get(),
```
Why not pass the correct category in here?

brentfulgham replied:
Good point -- I think after your suggested generator change there is never a case where this is unknown.

emw-apple commented on Source/WTF/Scripts/GeneratePreferences.rb:
```diff
@@ -266,6 +291,16 @@ def initializeParsedPreferences(parsedPreferences)
           next
         end
 
+        if %w{ developer testable preview stable }.include?(status)
```
Would we want to mandate a category for `internal` features, given that they are supposed to show up in internal UI?

patrickangle replied:
Since presenting internal features somewhere where categories could be used isn't an immediate concern, I think its okay if we punt of categorizing those settings.

brentfulgham replied:
I'll follow @patrickangle's lead here. We can easily adjust later as needed.

emw-apple commented on Source/WTF/Scripts/GeneratePreferences.rb:
```diff
@@ -96,6 +97,7 @@ def initialize(name, opts, frontend)
     @type = opts["type"]
     @refinedType = opts["refinedType"]
     @status = opts["status"]
+    @category = opts["category"]
```
nit: I think I'd prefer defaulting to "none" here instead of in the getter methods.
```suggestion
    @category = opts["category"] || "none"
```

brentfulgham replied:
Oh, that's a great idea -- I'll adopt that.

patrickangle commented on Source/WebKitLegacy/mac/WebView/WebFeature.h:
```diff
@@ -50,12 +50,28 @@ typedef NS_ENUM(NSUInteger, WebFeatureStatus) {
 
+/*!
+ @enum WebFeatureCategory
+ @abstract Field indicating the category of a web feature. Used to determine where a feature should be shown in the UI.
```
Instead of saying "where a feature should be shown in the UI", perhaps we mean "how the feature should be sorted or grouped in the UI"?

brentfulgham replied:
I like it -- will revise.

patrickangle reviewed (approved):
This looks good to me from the API end of things. I haven't deeply scrutinized which category each setting went into, but that's likely best handled as followup if folks want to debate what goes where.

brentfulgham committed:
Expose a 'category' for Web Preferences
https://bugs.webkit.org/show_bug.cgi?id=252093
<rdar://104962882>

Reviewed by Patrick Angle.

Add a new 'category' field to our preference definitions that can be used to provide context
to the User Interface for applications like Safari so that we can easily present the features
in the clearest way.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Scripts/GenerateSettings.rb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/APIFeature.cpp
(API::Feature::create): Add new argument.
(API::Feature::Feature): Ditto.
* Source/WebKit/UIProcess/API/APIFeature.h:
* Source/WebKit/UIProcess/API/APIFeatureStatus.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm:
(-[_WKFeature category]):
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesExperimentalFeatures.mm.erb:
* Source/WebKitLegacy/mac/Scripts/PreferencesTemplates/WebPreferencesInternalFeatures.mm.erb:
* Source/WebKitLegacy/mac/WebView/WebFeature.h:
* Source/WebKitLegacy/mac/WebView/WebFeature.m:
(-[WebFeature initWithKey:preferenceKey:name:status:category:details:defaultValue:hidden:]): Added.
(-[WebFeature initWithKey:preferenceKey:name:status:details:defaultValue:hidden:]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebFeatureInternal.h:

Canonical link: https://commits.webkit.org/260420@main
